src/System.Private.CoreLib/shared/System/String.Comparison.cs

```diff
@@ -787,6 +788,130 @@ internal unsafe int GetNonRandomizedHashCode()
+            {
+                uint hash1 = (5381 << 16) + 5381;
+                uint hash2 = hash1;
+
+                char* ptr = src;
+                int length = this.Length;
+
+                while (length >= 2)
+                {
+                    char value1 = ptr[0];
+                    char value2 = ptr[1];
+
+                    if ((value1 | value2) >= 0x80)
+                    {
+                        // Non-ascii detected, fallback to non-ascii aware casing
+                        return GetRemainingIgnoreCaseHashCode(ptr, length, hash1, hash2);
+                    }
+
+                    // If we have a uppercase character, ORing with 0x20
+                    // will make it an lowercase character.
+                    if ((uint)(value1 - 'A') <= (uint)('Z' - 'A'))
+                    {
+                        value1 = (char)(value1 | 0x20);
+                    }
+                    if ((uint)(value2 - 'A') <= (uint)('Z' - 'A'))
+                    {
+                        value2 = (char)(value2 | 0x20);
+                    }
+
+                    length -= 2;
+                    hash1 = (((hash1 << 5) | (hash1 >> 27)) + hash1) ^ value1;
+                    hash2 = (((hash2 << 5) | (hash2 >> 27)) + hash2) ^ value2;
+                    ptr += 2;
+                }
+
+                if (length > 0)
+                {
+                    char value = ptr[0];
+                    if (value >= 0x80)
+                    {
+                        // Non-ascii detected, fallback to non-ascii aware casing
+                        return GetRemainingIgnoreCaseHashCode(ptr, length, hash1, hash2);
+                    }
+
+                    // If we have a uppercase character, ORing with 0x20
+                    // will make it an lowercase character.
+                    if ((uint)(value - 'A') <= (uint)('Z' - 'A'))
+                    {
+                        value = (char)(value | 0x20);
+                    }
+
+                    hash2 = (((hash2 << 5) | (hash2 >> 27)) + hash2) ^ value;
+                }
+
+                return (int)(hash1 + (hash2 * 1566083941));
+            }
+        }
+
+        private static unsafe int GetRemainingIgnoreCaseHashCode(char* str, int remainingLength, uint hash1, uint hash2)
+        {
+            Debug.Assert(remainingLength > 0);
+            Debug.Assert(str[remainingLength] == '\0', "str[remainingLength] == '\\0'");
+
+            char[] borrowedArr = null;
+            Span<char> span = remainingLength <= 255 ?
+                stackalloc char[255] :
+                (borrowedArr = ArrayPool<char>.Shared.Rent(remainingLength));
+
+            int charsWritten = new ReadOnlySpan<char>(str, remainingLength).ToLowerInvariant(span);
+            span = span.Slice(0, charsWritten);
+
+            fixed (char* src = span)
+            {
+                str = src;
+                int length = span.Length;
+
+                if (((int)str & 3) != 0)
```

Comment:
Do I need to check `int` alignment here? Could be `stackalloc char[]` or `char[]`

Reply:
The address of the first character of a char[] definitely is int-aligned (the length is always right before it and it is an int).   

stackalloc is built on the localloc instruction and 
https://msdn.microsoft.com/en-us/library/system.reflection.emit.opcodes.localloc(v=vs.110).aspx specifies that it is uIntPtr aligned.     I would feel reasonably comfortable replacing the if with a Assert.

Reply:
Changed to Assert :)